src/tty.c

```diff
@@ -40,7 +40,13 @@ METAL_CONSTRUCTOR(metal_tty_init) {
  * provide a shim that eats all the characters so we can ensure that everything
  * can link to metal_tty_putc. */
 int nop_putc(int c) __attribute__((section(".text.metal.nop.putc")));
-int nop_putc(int c) { return -1; }
+// Use a customizable NOP hint instruction so that a post-processor parser can
+// look for this instruction, and use the value in a0 as the character to be
+// printed.
+int nop_putc(int c) {
+    __asm__ volatile("slli x0,a0,0x11");
```

Comment:
Now that I think about it, it's probably better to do an extended assembly instruction like

```
__asm__ volatile("slli x0, %0, 0x11" :: "r" (c));
```

Technically we can expect the ABI to put `c` in register `a0`, but this tells the compiler exactly what we want to happen.

Reply:
lgtm.  I [used a different nop ](https://github.com/sifive/freedom-metal/blob/30fd755314b578c5dba8196c14533619fde261b1/src/tty.c#L44)in the perf-tools branch of freedom-metal, but never PR'd it because it felt inappropriate, but if this is accepted, I'll gladly switch to it and make my branch unnecessary.

Reply:
Without @nategraff-sifive's fix, LTO could miscompile the code, since the compiler otherwise wouldn't know the dependence between `c` and the `asm` block.

Further constraining that rs1 be a0 would make the sed script easier to write, but that's a judgment call.

Reply:
Ah, okay, LTO is sufficient reason to tweak it. I'll do the silly move-to-self trick then, with a comment about working even if LTO would otherwise mess it up.

Reply:
@ericlove that `andi` nop is unfortunately in the nop hint space that is "reserved for future standard use." That's one minor reason why I went with the `slli` instead, so that it won't break when we/someone decide `andi x0, *` is some sort of hint.